### PR TITLE
Validation

### PR DESCRIPTION
**Pull request by svgogin**
1) Validation implemented
2) Questions written

### CODE REVIEW

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java [hunk 1]
```diff
@@ -35,4 +38,32 @@
...
+
+  @NonNull
+  @Override
+  protected ResponseEntity<Object> handleMethodArgumentNotValid(MethodArgumentNotValidException ex,
+                                                                @NonNull HttpHeaders headers,
+                                                                @NonNull HttpStatus status,
+                                                                @NonNull WebRequest request) {
+    var errorField = ex.getBindingResult().getFieldErrors().iterator().next().getField();
+    var message = "Invalid format of request parameter: " + errorField;
+    //toDO If a parameter is not passed in the request, I don't know,
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]: please, add tests and also for this case

File: src/main/java/ru/svgogin/service/spark/service/SparkService.java [hunk 2]
```diff
@@ -50,7 +50,7 @@ public CompanyDto update(String inn, CompanyDto companyDto) {
                                                      + " doesn't exist"));
     Company companyForUpdate = actualize(companyDtoEx, companyDto);
     var updatedCompany = aggregateTemplate.update(companyForUpdate);
-    log.info("Company with inn " + inn + " was successfully updated.");
+    log.info("Company with inn {}} was successfully updated.", inn);
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/service/SparkService.java]:
```suggestion
    log.info("Company with inn {} was successfully updated.", inn);
```

File: src/main/java/ru/svgogin/service/spark/web/SparkController.java [hunk 3]
```diff
@@ -33,26 +36,36 @@ public SparkController(SparkService sparkService) {
...
     log.info("updateCompany " + companyDto.getInn());
     return new ResponseEntity<>(sparkService.update(inn, companyDto), HttpStatus.OK);
   }
 
   @DeleteMapping("/{inn}")
-  public ResponseEntity<CompanyDto> deleteCompany(@PathVariable("inn") String inn) {
+  public ResponseEntity<CompanyDto> deleteCompany(@PathVariable("inn")
+                                                    @Pattern(regexp = "^(\\d{10}|\\d{12})$",
+                                                             message = "Inn should contain "
+                                                        + "10 or 12 digits") String inn) {
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/web/SparkController.java]: It is better to format this way (and in another methods)
```suggestion
  public ResponseEntity<CompanyDto> deleteCompany(
      @PathVariable("inn")
      @Pattern(regexp = "^(\\d{10}|\\d{12})$", message = "Inn should contain 10 or 12 digits") String inn
      ) {
```

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java [hunk 4]
```diff
@@ -35,4 +41,37 @@
...
+        new HttpHeaders(), HttpStatus.BAD_REQUEST, request);
+  }
+
+  @NonNull
+  @Override
+  protected ResponseEntity<Object> handleMethodArgumentNotValid(MethodArgumentNotValidException ex,
+                                                                @NonNull HttpHeaders headers,
+                                                                @NonNull HttpStatus status,
+                                                                @NonNull WebRequest request) {
+    var error = ex.getBindingResult().getFieldErrors().iterator().next();
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]: why do you return error only for first field?

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java [hunk 5]
```diff
@@ -35,4 +41,37 @@
...
+    var errorField = error.getField();
+    //This is a bullshit implementation, but I have no idea, how to improve it
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]: it is better to do it this way:
```java
  protected ResponseEntity<Object> handleMethodArgumentNotValid(MethodArgumentNotValidException ex,
                                                                @NonNull HttpHeaders headers,
                                                                @NonNull HttpStatus status,
                                                                @NonNull WebRequest request) {
    var error = ex.getBindingResult().getFieldErrors().iterator().next();
    var errorField = error.getField();
    var errorCode = errorCodeBy(error);
    var message = String.format("%s (%s)", errorCode.message, errorField);
    log.warn(message);
    var bodyOfResponse = new ErrorDto(errorCode, message);
    return handleExceptionInternal(
        ex, bodyOfResponse, headers, status, request
    );
  }

  private ErrorCode errorCodeBy(FieldError fieldError) {
    var notBlankExists = Optional.ofNullable(fieldError.getCodes())
        .map(codes -> Set.of(codes).contains("NotBlank"))
        .orElse(false);
    if (notBlankExists) {
      return ErrorCode.ERROR004;
    } else {
      return ErrorCode.ERROR003;
    }
  }
```
But also you shouldn't return result only for first error

File: src/test/java/ru/svgogin/service/spark/web/SparkControllerItTest.java [hunk 6]
```diff
@@ -249,8 +249,169 @@ void saveCompanyShouldThrowExceptionWhenExists() throws Exception {
...
                 "{\n"
                 + "    \"code\": \"ERROR001\",\n"
-                + "    \"message\": \"Company with inn 07725038124 already exists\"\n"
+                + "    \"message\": \"Company with inn 7725038124 already exists\"\n"
                 + "}\n"
         ));
   }
+
+  @Test
+  void getCompanyByInnShouldThrowException_WhenPathInnIsNotValid() throws Exception {
```
Comment (alinkov) [inline src/test/java/ru/svgogin/service/spark/web/SparkControllerItTest.java]: why do you test it in integration test?

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java (hunk 1 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]: Done

File: src/main/java/ru/svgogin/service/spark/service/SparkService.java (hunk 2 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/service/SparkService.java]: Done

File: src/main/java/ru/svgogin/service/spark/web/SparkController.java (hunk 3 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/web/SparkController.java]: [INFO] --- maven-checkstyle-plugin:3.1.1:check (validate) @ spark ---
[INFO] Starting audit...
[ERROR] D:\ServiceRepository\src\main\java\ru\svgogin\service\spark\web\SparkController.java:66: Line is longer than 100 characters (found 154). [LineLength]
Audit done.

Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/web/SparkController.java]: Done

File: src/test/java/ru/svgogin/service/spark/web/SparkControllerItTest.java (hunk 6 above)
Comment (svgogin) [reply src/test/java/ru/svgogin/service/spark/web/SparkControllerItTest.java]: Tests have been moved to SparkControllerTest

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java (hunk 5 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]: Done

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java [hunk 7]
```diff
@@ -1,38 +1,86 @@
...
 import ru.svgogin.service.spark.errordto.ErrorDto;
+import ru.svgogin.service.spark.errordto.ErrorDto.ErrorCode;
 import ru.svgogin.service.spark.exception.EntityAlreadyExistsException;
 import ru.svgogin.service.spark.exception.NoSuchEntityException;
 
 @ControllerAdvice
 public class RestResponseEntityExceptionHandler extends ResponseEntityExceptionHandler {
   private static final Logger log = LoggerFactory
       .getLogger(RestResponseEntityExceptionHandler.class);
+  ErrorCode errorCode;
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]: why do you this?
This class is a singleton and can work in multiple threads. In this case you can return incorrect result

File: src/test/java/ru/svgogin/service/spark/web/SparkControllerTest.java [hunk 8]
```diff
@@ -198,20 +198,184 @@ void deleteCompanyShouldDeleteEntities() throws Exception {
...
     //when
-    mockMvc.perform(MockMvcRequestBuilders.delete("/spark/companies/97725038124"))
+    mockMvc.perform(MockMvcRequestBuilders.delete("/spark/companies/9772503812"))
         .andDo(print())
         .andExpect(MockMvcResultMatchers.status().isNotFound());
   }
+
+  @Test
+  void getCompanyByInnShouldThrowException_WhenPathInnIsNotValid() throws Exception {
+    // given
```
Comment (alinkov) [inline src/test/java/ru/svgogin/service/spark/web/SparkControllerTest.java]:
```suggestion
```
nit: You can remove this line

File: src/test/java/ru/svgogin/service/spark/web/SparkControllerTest.java [hunk 9]
```diff
@@ -198,20 +198,184 @@ void deleteCompanyShouldDeleteEntities() throws Exception {
...
+            "[{\n"
+            + "    \"code\": \"ERROR003\",\n"
+            + "    \"message\": \"getCompanyByInn.inn: Inn should contain 10 or 12 digits\"\n"
+            + "}]\n"
+            )
+        );
+  }
+
+  @Test
+  void saveCompanyShouldThrowException_WhenOgrnInBodyIsNotValid() throws Exception {
```
Comment (alinkov) [inline src/test/java/ru/svgogin/service/spark/web/SparkControllerTest.java:234]:
```suggestion
  void saveCompanyShouldThrowException_WhenOgrnIsNotValid() throws Exception {
```

And also, you can use snake_case for tests:
```suggestion
  void save_company_should_throw_exception_when_ogrn_is_not_valid() throws Exception {
```

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java (hunk 7 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]: Fixed

File: src/test/java/ru/svgogin/service/spark/web/SparkControllerTest.java (hunk 9 above)
Comment (svgogin) [reply src/test/java/ru/svgogin/service/spark/web/SparkControllerTest.java:234]: Ok

File: src/test/java/ru/svgogin/service/spark/web/SparkControllerTest.java (hunk 8 above)
Comment (svgogin) [reply src/test/java/ru/svgogin/service/spark/web/SparkControllerTest.java]: Done

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java [hunk 10]
```diff
@@ -20,19 +31,66 @@
...
     return handleExceptionInternal(ex, bodyOfResponse,
         new HttpHeaders(), HttpStatus.CONFLICT, request);
   }
 
   @ExceptionHandler(value = {NoSuchEntityException.class})
   protected ResponseEntity<Object> handleNotFound(RuntimeException ex, WebRequest request) {
-    String message = ex.getMessage();
+    var message = ex.getMessage();
+    var errorCode = ErrorCode.ERROR002;
+    var bodyOfResponse = List.of(new ErrorDto(errorCode, message));
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java:46]: nit: you can inline variables:
```suggestion
    var bodyOfResponse = List.of(new ErrorDto(ErrorCode.ERROR002, ex.getMessage()));
```

**Review by alinkov — APPROVED** [review]

**Pull request merged**